storage/storage_image.go

```diff
@@ -360,6 +367,57 @@ func (s *storageImageDestination) HasThreadSafePutBlob() bool {
 	return true
 }
 
+// SupportsBlobLocks indicates whether the ImageDestination supports blob
+// locking.
+func (s *storageImageDestination) SupportsBlobLocks() bool {
+	return true
+}
+
+// LockBlob can be used to synchronize operations on it (e.g., copying).
+// Note that any unlocked lock will be unlocked in Close().
+func (s *storageImageDestination) LockBlob(b types.BlobInfo) error {
+	// anonymous function to not worry about errors paths and panics
+	getLocker := func() (storage.Locker, error) {
+		s.digestLockMutex.Lock()
+		defer s.digestLockMutex.Unlock()
+		if locker, ok := s.lockedDigests[b.Digest]; ok {
+			return locker, nil
+		}
+		locker, err := s.imageRef.transport.store.GetDigestLock(b.Digest)
+		if err != nil {
+			return locker, err
+		}
+		s.lockedDigests[b.Digest] = locker
+		return locker, nil
+	}
+
+	locker, err := getLocker()
+	if err != nil {
```

Comment:
FIXME:

At this point, `s.digestLockMutex` is unlocked; so, this thread can be suspended, and some other thread can concurrently call _both_ `LockBlob` and `UnlockBlob` before this thread resumes; that `UnlockBlob` will remove `locker` from `s.lockedDigests`; then this thread could resume, call `locker.Lock()`, and we will end up with a locked `locker` that is not present in `s.lockedDigests`.

Reply:
Good catch!

I solved it by removing the entries from `s.lockedDigests` only in `Close()`.

Reply:
(no further updates to the FIXME.)

Reply:
Update: we leave deletion up to the garbage collector and use a 2nd map `unlockedDigest` that keeps track of the state.